### DISCUSSION TITLE
Is it possible to use a generalized library with weak sindy?

### DISCUSSION THREAD

**Discussion by scatr** [dynamicslab/pysindy#389] [General]
Hi,

I have a PDE model which features quite a complicated set of features, but I want to regress only a handful of unknown coefficients on to the exact known model. 

If I define a general library which contains all the different terms in the PDE, and allows for all inputs, this results in a library with lots of terms and constraining the library explicitly after construction becomes too difficult as there are 1000s of terms. I really only need 10 or so in each equation. 

It would be easier to limit the number of terms in each equation first using the generalized Library. So for instance, a term of the form x*y**0.5 would appear with the correct inputs only. 

Is it possible to use the generalized library feature with the weak formulation? Or is there another method for constructing the exact known model? 

Is it possible to construct different libraries for each equation? 

Thanks in advance

**Comment by Jacob-Stevens-Haas** [comment]
yes, there is - The `GeneralizedLibrary` contains a sequence of libraries, and the `inputs_per_library` argument allows you to specify which inputs go in which feature library.  See #306 for an explanation of the old semantics (most recent release) and the current semantics (master branch).

**Reply by scatr** [reply]
Thanks for your reply, would you be able to check that this makes sense for one equation of the problem? I have one equation for the problem given as 
$$\partial_t R_{11}  + u_z \partial_z R_{11} + 2 R_{12} \partial_z w - \nu \partial_{zz} R_{11} = k_1 R^{1/2} R_{11} + k_2R^{3/2} - k_3 R_{11}$$ 
where the LHS of the problem contains known coefficients and $k_1, k_2, k3$ are the unknowns I need to find. The data state vector contains the variables 
$$x = [u,w,R, R_{11}, R_{12}, R_{22}, Fx, Fz, Q]$$.
forming the LHS of my regression problem. Not all of the variables were used in this particular equation. To create the library for this particular equation using the `GeneralizedLibrary` feature you mentioned, I have used the following code:

```
feature_names = ["Ubar", "Wbar", "R", "R11", "R12", "R22", "Fx", "Fz", "Q"]

library_functions_linear = [lambda x: x] 
library_functions_sqrt = [lambda x,y: x*np.abs(y)**(1/2)] 
library_functions_32 = [lambda x: np.abs(x)**(3/2)]
library_functions_quad = [lambda x,y: x*y]
library_names_linear = [lambda x: x]
library_names_sqrt = [lambda x,y: x+y+"^1/2"]
library_names_32 = [lambda x: x+"^3/2"]
library_names_quad = [lambda x,y: x+y]

#create the R11 and R11_zz terms
weak_lin_lib = ps.WeakPDELibrary(
    library_functions=library_functions_linear,
    function_names = library_names_linear,
    derivative_order=2, 
    spatiotemporal_grid=XT,
    include_interaction = False, #remove the nonlinear terms this generates
    K=2000,
    H_xt=[Lz / 10, (end-start) / 10],
)

#Create the R^(1/2)R11 terms
weak_quad_lib = ps.WeakPDELibrary(
    library_functions=library_functions_sqrt,
    function_names = library_names_sqrt,
    derivative_order=0,
    spatiotemporal_grid=XT,
    K=2000,
    H_xt=[Lz / 10, (end-start) / 10],
)

#create the Rbar^(3/2) term 
weak_32_lib = ps.WeakPDELibrary(
    library_functions=library_functions_32,
    function_names = library_names_32,
    derivative_order=0,
    spatiotemporal_grid=XT,
    K=2000,
    H_xt=[Lz / 10, (end-start) / 10],
)

#create R12 wbar_z
nonlinear_1_lib = ps.WeakPDELibrary(
    library_functions=library_functions_linear,
    function_names = library_names_linear,
    derivative_order=1,
    spatiotemporal_grid=XT,
    include_interaction = True, 
    K=2000,
    H_xt=[Lz / 10, (end-start) / 10],
)


#create ubar_z R11_z
nonlinear_2_lib = ps.WeakPDELibrary(
    library_functions=library_functions_linear,
    function_names = library_names_linear,
    derivative_order=1,
    spatiotemporal_grid=XT,
    include_interaction = True, 
    K=2000,
    H_xt=[Lz / 10, (end-start) / 10],
)

generalized_lib = ps.GeneralizedLibrary(
                    [weak_lin_lib, weak_quad_lib, weak_32_lib, nonlinear_1_lib,
                     nonlinear_2_lib],
                    inputs_per_library = [[3],[2,3],[2], [1, 4], [0, 3]]) 

# Fit the clean data
optimizer = ps.STLSQ(threshold=5e-6, alpha=1e-12, normalize_columns=False)
model = ps.SINDy(
    feature_library=generalized_lib, optimizer=optimizer, feature_names = feature_names
)
model.fit(u_shaped_noiseless, t=t, unbias=False)
print("Noiseless weak fit:")
model.print()
print("R2 score: ", model.score(u_shaped_noiseless, t=t))

print("The terms included in our fitting library for the R11 equation are:")
model.get_feature_names()
```

which gives the library terms as 

``` 
['R11',
 'R11_1',
 'R11_11',
 'RR11^1/2',
 'R^3/2',
 'Wbar',
 'R12',
 'Wbar_1',
 'R12_1',
 'WbarWbar_1',
 'R12Wbar_1',
 'WbarR12_1',
 'R12R12_1',
 'Ubar',
 'R11',
 'Ubar_1',
 'R11_1',
 'UbarUbar_1',
 'R11Ubar_1',
 'UbarR11_1',
 'R11R11_1']
```
This seems to give me a manageable number of terms which I can then constrain with a different optimizer like `ConstrainedSR3`. Is this the correct usage? I was also unsure if the term $R_{12} \partial_z w$ presented issues finding the weak form? Many thanks for any advice.
